tests/app/main/test_suppliers.py

```diff
@@ -2322,3 +2324,188 @@ def test_post_shows_already_entered_page_and_api_not_called_if_data_already_ente
+    ):
+        with self.app.test_client():
+            self.data_api_client.get_supplier.return_value = get_supplier()
+
+            res = getattr(self.client, method)("/suppliers/vat-number/edit")
+            doc = html.fromstring(res.get_data(as_text=True))
+
+            assert res.status_code == status_code
+            assert doc.xpath("//h1[normalize-space(string())='Change your VAT number']")
+            assert not self.data_api_client.update_supplier.called
+
+    def test_posting_valid_vat_code_updates_supplier(self):
+        with self.app.test_client():
+            res = self.client.post(
+                "/suppliers/vat-number/edit",
+                data={
+                    "vat_registered": "Yes",
+                    "vat_number": "GB123456789",
+                }
+            )
+
+            assert res.status_code == 302
+            assert res.location == "http://localhost/suppliers/details"
+            self.data_api_client.update_supplier.assert_called_once_with(
+                supplier_id=1234,
+                supplier={"vatNumber": "GB123456789"},
+                user="email@email.com"
+            )
+
+    @pytest.mark.parametrize('vat_number', ("", "Not a VAT number"))
+    def test_posting_not_vat_registered_correctly_updates_supplier(self, vat_number):
+        with self.app.test_client():
+            res = self.client.post(
+                "/suppliers/vat-number/edit",
+                data={
+                    "vat_registered": "No",
+                    "vat_number": vat_number,
+                }
+            )
+
+            assert res.status_code == 302
+            assert res.location == "http://localhost/suppliers/details"
+            self.data_api_client.update_supplier.assert_called_once_with(
+                supplier_id=1234,
+                supplier={"vatNumber": "Not VAT registered"},
+                user="email@email.com",
+            )
+
+    @pytest.mark.parametrize("vat_number, message, masthead", (
+        ("GB123456789", None, None),
+        ("GB123456789012", None, None),
+        ("GBGD123", None, None),
+        ("GBHA456", None, None),
+        ("123456789", None, None),
+        ("123456789012", None, None),
+        ("GD123", None, None),
+        ("HA456", None, None),
+        ("NOTVALID", "You must provide a valid VAT number, they are usually either 9 or 12 digits.", "VAT number"),
+        ("", "You must provide a VAT number.", "VAT number")
+    ))
+    def test_validation_of_tax_number(self, vat_number, message, masthead):
+        with self.app.test_client():
+            res = self.client.post(
+                "/suppliers/vat-number/edit",
+                data={
+                    "vat_registered": "Yes",
+                    "vat_number": vat_number,
+                }
+            )
+            doc = html.fromstring(res.get_data(as_text=True))
+
+            if message:
+                assert res.status_code == 400
+                assert doc.xpath(
```

Comment:
We have a helper method to validate mastheads now, `assert_single_question_page_validation_errors`. It would be good to use that here.